### PR TITLE
Added the scorecard github action and its badge

### PR DESCRIPTION
**Pull request by harshitasao**
## Description

PR to add the [Scorecard GitHub Action](https://github.com/ossf/scorecard-action) and its [badge](https://openssf.org/blog/2022/09/08/show-off-your-security-score-announcing-scorecards-badges/) in the README file.

## Related Issue(s)
  - Fixes: #16330 

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [ ] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [ ] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on CI?
-   [ ] Documentation was added or is not required

### CODE REVIEW

File: .github/workflows/scorecards.yml
Comment (mattlord) [inline .github/workflows/scorecards.yml:9]: Any particular reason for this value? It would be nice to add a comment ~ `# Every Wednesday at 9:45AM`

File: .github/workflows/scorecards.yml
Comment (mattlord) [inline .github/workflows/scorecards.yml:11]: We do many pushes (merging a PR) to main per week. Can we then do without the scheduled job?

File: .github/workflows/scorecards.yml
Comment (mattlord) [inline .github/workflows/scorecards.yml:19]: Do we expect this to be fairly resource intensive? If so, we have access to larger runners via CNCF.

File: .github/workflows/scorecards.yml
Comment (mattlord) [inline .github/workflows/scorecards.yml:31]: Is there any reason not to use `actions/checkout@v4`?

File: .github/workflows/scorecards.yml
Comment (mattlord) [inline .github/workflows/scorecards.yml:36]: Any reason not to use `ossf/scorecard-action@v2` so that the version is more clear and we automatically use the latest version available?

File: .github/workflows/scorecards.yml
Comment (mattlord) [inline .github/workflows/scorecards.yml:58]: Any reason not to use `actions/upload-artifact@v3` or similar?

**Review by mattlord — APPROVED** [review]
Thanks, @harshitasao! I only had a few minor questions and suggestions. I'll approve since we still need a second approver before merging anyhow.

File: .github/workflows/scorecards.yml
Comment (dbussink) [reply .github/workflows/scorecards.yml:9]: I'd very much recommend we don't pick a time that's a specific human nice round number thing like "45" minutes. It's when there's many other crons too. Better to pick something actually random instead :smile:. 

But I also agree with @mattlord's later comment, do we need a cron at all? Is it too expensive to run on each push to main?

File: .github/workflows/scorecards.yml
Comment (dbussink) [reply .github/workflows/scorecards.yml:31]: Yeah, let's not ping specific shas here but the latest semantic version.

File: .github/workflows/scorecards.yml
Comment (harshitasao) [reply .github/workflows/scorecards.yml:9]: @mattlord  @dbussink  The time is generated by the scorecard github action. 
The action itself is not expensive, but running this action on every push to the main may make it expensive.

File: .github/workflows/scorecards.yml
Comment (harshitasao) [reply .github/workflows/scorecards.yml:19]: Sorry, I didn't understand your question. Could you please clarify.

File: .github/workflows/scorecards.yml
Comment (harshitasao) [reply .github/workflows/scorecards.yml:31]: IMO I think we should not change this as in the scorecard action there is a check for [pinned dependencies](https://github.com/ossf/scorecard/blob/cfe5c5f9c3dbf2a98c864e1b1ea25fae289f3b9d/docs/checks.md#pinned-dependencies) (this project has many workflows which are not pinned by hash resulting score of 0 which needs improvement) so if we change this it will further drop the score.

Comment (dbussink) [reply .github/workflows/scorecards.yml:31]: We should not pin dependencies imho without also added automation to update it. Since otherwise it never gets updated.

Comment (harshitasao) [reply .github/workflows/scorecards.yml:31]: I assumed that the project already has dependabot, Sorry for that, I will make the changes. Also I suggest adding Dependency-Update-Tool to the project; it will be a good addition, as [Dependency-Update-Tool](https://github.com/ossf/scorecard/blob/a8eae2d833553e93021a22f5641661cb6ad3e24c/docs/checks.md#dependency-update-tool) is also one of the scorecard checks, and most of the CNCF projects use it. With the addition of the dependency update tool, we will be able to pin dependencies, which will improve the scorecard score.

File: .github/workflows/scorecards.yml
Comment (GrahamCampbell) [reply .github/workflows/scorecards.yml:36]: The reason is OSS supply chain vulnerabilities, however pinning a particular hash is still dubious protection if there is not detailed auditing when a version upgrade is explicitly applied.

Comment (GrahamCampbell) [reply .github/workflows/scorecards.yml:36]: https://docs.github.com/en/actions/security-for-github-actions/security-guides/security-hardening-for-github-actions#using-third-party-actions

File: .github/workflows/scorecards.yml
Comment (deepthi) [reply .github/workflows/scorecards.yml:31]: We do have dependabot running on the project. However, I have not seen it update GH workflows, and in some other context we found that bots are not allowed to update GH workflows.

File: .github/workflows/scorecards.yml
Comment (deepthi) [reply .github/workflows/scorecards.yml:19]: ubuntu-latest is the default GH actions runner. Since you said that the action itself is not expensive, this is probably fine.

File: .github/workflows/scorecards.yml
Comment (deepthi) [reply .github/workflows/scorecards.yml:31]: bot PRs: https://github.com/vitessio/vitess/pulls?q=is%3Apr+author%3Aapp%2Fdependabot+is%3Aclosed

**Comment by deepthi** [comment]
@harshitasao it makes sense to get this reviewed by one of your mentors as well.

**Comment by harshitasao** [comment]
> @harshitasao it makes sense to get this reviewed by one of your mentors as well.

Sorry @deepthi I missed this message.

@joycebrum @diogoteles08 @pnacht Could you please review this PR.

**Review by pnacht — APPROVED** [review]

File: .github/workflows/scorecards.yml
Comment (frouioui) [reply .github/workflows/scorecards.yml:11]: Coming back to this @harshitasao, any reason why we would not want to remove the `schedule` trigger and only keep branch protection rules and main?

Comment (deepthi) [reply .github/workflows/scorecards.yml:11]: Agree with others on this. Let's remove the schedule, and run it on `push`.

**Comment by frouioui** [comment]
@harshitasao would you mind merging `main` on top of your branch? This will allow tests to go green.

**Review by deepthi — APPROVED** [review]
Once CI is green we can go ahead and merge.

**Review by deepthi — APPROVED** [review]

**Comment by deepthi** [comment]
Thank you @harshitasao!

**Comment by mattlord** [comment]
This seems to be failing on every push to main: https://github.com/vitessio/vitess/commits/main/

The same error each time:
```
Error: Unable to resolve action `ossf/scorecard-action@v2`, unable to find version `v2`
```

**Comment by pnacht** [comment]
Hey everyone. The issue here is that the Scorecard Action doesn't have tags for major versions, only patch versions. So it needs to be 

```
uses: ossf/scorecard-action@v2.4.0
```

Sorry, this went unnoticed when I reviewed the code before it got merged.

**Comment by mattlord** [comment]
> Hey everyone. The issue here is that the Scorecard Action doesn't have tags for major versions, only patch versions.

Yeah, I saw that: https://github.com/ossf/scorecard-action/tags

> Sorry, this went unnoticed when I reviewed the code before it got merged.

No worries! We'll get it updated soon. ❤️

**Comment by mattlord** [comment]
Still failed on the first run after addressing the action version: 
```
Using payload from: results.json
Generating ephemeral keys...
Retrieving signed certificate...
2024/09/08 19:15:09 error signing scorecard results: getting key from Fulcio: retrieving cert: POST https://fulcio.sigstore.dev/api/v1/signingCert returned 502 Bad Gateway: "<!doctype html><meta charset=\"utf-8\"><meta name=viewport content=\"width=device-width, initial-scale=1\"><title>502</title>502 Bad Gateway"
2024/09/08 19:15:09 retrying in 10s...
2024/09/08 19:15:19 error signing scorecard json results: error signing payload: getting key from Fulcio: retrieving cert: POST https://fulcio.sigstore.dev/api/v1/signingCert returned 502 Bad Gateway: "<!doctype html><meta charset=\"utf-8\"><meta name=viewport content=\"width=device-width, initial-scale=1\"><title>502</title>502 Bad Gateway"
```

Hopefully that's an ephemeral issue though.

**Comment by pnacht** [comment]
I just checked with the Fulcio team (a Scorecard dependency to publish the results). They were down for some maintenance over the weekend.

Could you try rerunning the workflow? It should work now.

**Pull request merged**
